### PR TITLE
Update edition template with optional limit, manual closing, burn tracking

### PR DESCRIPTION
**Pull request by psiemens**
This PR proposes changes to the standard edition template that enable the following:

- **Open editions:** developers can now create editions without a predefined limit.
- **Manual closing:** developers can now close an edition at any time.
  - This emits an `EditionClosed` event (which is also automatically emitted when an edition reaches its limit).
- **Burn tracking:** editions now track the number of NFTs that have been burned.

I'm looking for feedback on this change!

### CODE REVIEW

File: packages/core/contracts/EditionNFTContract.test.ts [hunk 1]
```diff
@@ -54,125 +54,310 @@ describe('EditionNFTContract', () => {
...
+      expect(onChainEdition.size).toEqual(count);
+    });
 
-  it('should fail to mint more than edition size', async () => {
-    await expect(async () => {
-      await client.send(contract.mintNFTs({ editionId: edition1.id, count: 5 }));
-    }).rejects.toThrow();
+    it('should fail to mint after closing', async () => {
+      await expect(async () => {
+        await client.send(contract.mintNFTs({ editionId: edition.id, count: 5 }));
```
Comment (jacksonConrad) [inline packages/core/contracts/EditionNFTContract.test.ts:216]: Test would be cleaner if you attempt to mint 1 NFT.  Since the limit is 5 and the size here is 3, calling `mintNFTs` with `count: 5` could be erroring because you would theoretically exceed the limit, rather than because the edition is closed.
```suggestion
        await client.send(contract.mintNFTs({ editionId: edition.id, count: 1 }));
```

File: packages/core/contracts/EditionNFTContract.test.ts [hunk 2]
```diff
@@ -54,125 +54,310 @@ describe('EditionNFTContract', () => {
...
 
-    expect(onChainEdition.id).toEqual(edition1.id);
-    expect(onChainEdition.size).toEqual(edition1.size);
-    expect(onChainEdition.count).toEqual(edition1.size);
-  });
+      expect(onChainEdition.id).toEqual(edition.id);
+      expect(onChainEdition.limit).toEqual(edition.limit);
 
-  const edition2Bucket = 'edition2';
+      // Size should still be the limit
```
Comment (jacksonConrad) [inline packages/core/contracts/EditionNFTContract.test.ts:224]:
```suggestion
      // Size should still be the count
```

File: packages/core/contracts/EditionNFTContract.test.ts [hunk 3]
```diff
@@ -54,125 +54,310 @@ describe('EditionNFTContract', () => {
...
+      await expect(async () => {
+        await client.send(contract.mintNFTs({ editionId: edition.id, count: 5 }));
+      }).rejects.toThrow();
+
+      const onChainEdition = await client.query(contract.getEdition(edition.id));
+
+      expect(onChainEdition.id).toEqual(edition.id);
+      expect(onChainEdition.limit).toBeUndefined();
+
+      // Size should still be the limit
```
Comment (jacksonConrad) [inline packages/core/contracts/EditionNFTContract.test.ts:285]:
```suggestion
      // Size should still be the count
```

**Review by jacksonConrad — APPROVED** [review]
LGTM!  just recommended some small comment/test tweaks

File: packages/core/contracts/EditionNFTContract.test.ts (hunk 1 above)
Comment (psiemens) [reply packages/core/contracts/EditionNFTContract.test.ts:216]: Great catch, thank you!

File: cadence/nfts/edition-nft/EditionNFT.template.cdc [hunk 4]
```diff
@@ -236,22 +301,31 @@ pub contract {{ contractName }}: NonFungibleToken {
             let edition = {{ contractName }}.editions[editionID]
                 ?? panic("edition does not exist")
 
-            // Increase the edition count by one
-            edition.incrementCount()
+            // Increase the edition size by one
+            edition.incrementSize()
```
Comment (fkenji) [inline cadence/nfts/edition-nft/EditionNFT.template.cdc:305]: Do we need a check / assert here to verify if the edition size has reached its limit? 

```
         if let limit = edition.limit {
                if edition.size == limit {
                    edition.close()
```

Wouldn't this line let the mint to go through?

**Review by fkenji — COMMENTED** [review]
Added some comments, let me know what you think 👀

File: cadence/nfts/edition-nft/EditionNFT.template.cdc [hunk 5]
```diff
@@ -224,6 +272,23 @@ pub contract {{ contractName }}: NonFungibleToken {
...
+        pub fun closeEdition(editionID: UInt64) {
+            let edition = {{ contractName }}.editions[editionID]
+                ?? panic("edition does not exist")
+
+            edition.close()
+
+            // Save the updated edition
+            {{ contractName }}.editions[editionID] = edition
+
+            emit EditionClosed(id: edition.id, size: edition.size)
```
[the remark this answers is not in the document]
Comment (psiemens) [reply cadence/nfts/edition-nft/EditionNFT.template.cdc:289]: I think it's OK to leave it out. The `limit` is a constant and is emitted in the `EditionCreated` event, so an event consumer should already have it.

File: cadence/nfts/edition-nft/EditionNFT.template.cdc (hunk 4 above)
Comment (psiemens) [reply cadence/nfts/edition-nft/EditionNFT.template.cdc:305]: Thanks for pointing this out. I confirmed that the logic works correctly, but it's definitely confusing.

There was a precondition in `incrementSize` that prevented the caller from increasing the size of a closed edition. Given that an edition is always closed when `size == limit`, it was impossible to increase the size past the limit. However, I updated the code to hopefully make that enforcement more clear.

- I adjusted the test cases to specifically check this edge case (minting 1 more than the limit)
- I moved the `isClosed` check from the `Edition` struct into the `mintNFT` function. This way it's easier to see all the logic in one function.

**Review by fkenji — APPROVED** [review]
Awesome! 🎉

**Comment by psiemens** [comment]
> LGTM, do we want a reopen functionality or is that out of scope for this?

Out of scope, but could be added later. As of now, closing an edition is meant to be a final action.
